### PR TITLE
revert old thread order logic

### LINKED ISSUE #986: [Bug] Changes made in commit #81ee52d mean that register order is no longer maintained in `context`
### GEF+GDB version

```shell
master branch of GEF
```

### Operating System

Ubuntu, Debian

### Describe the issue you encountered

Previously list comprehensions were used for filtering the registers requested in `DetailRegistersCommand(GenericCommand)` (around line 6733). This got changed in commit 81ee52d to use `set` instead.

Because `set` is unordered, changing the way the register filtering is done results in the registers printing in an unexpected order.

Previously you'd get `$rax`, `$rbx`, `$rcx` in order:

```
$rax   : 0x0               
$rbx   : 0x0               
$rcx   : 0x0               
$rdx   : 0x0               
$rsp   : 0x00007fffffffde70  →  0x0000000000000001
$rbp   : 0x0               
$rsi   : 0x0               
$rdi   : 0x0               
$rip   : 0x00007ffff7fe32b0  →  <_start+0> mov rdi, rsp
$r8    : 0x0               
$r9    : 0x0               
$r10   : 0x0               
$r11   : 0x0               
$r12   : 0x0               
$r13   : 0x0               
$r14   : 0x0               
$r15   : 0x0               
$eflags: [zero carry parity adjust sign trap INTERRUPT direction overflow resume virtualx86 identification]
$cs: 0x0033 $ss: 0x002b $ds: 0x0000 $es: 0x0000 $fs: 0x0000 $gs: 0x0000 
```

Now you get:

```
$rdx   : 0x0               
$rsi   : 0x0               
$rdi   : 0x0               
$rbp   : 0x0               
$eflags: [zero carry parity adjust sign trap INTERRUPT direction overflow resume virtualx86 identification]
$r8    : 0x0               
$rbx   : 0x0               
$r12   : 0x0               
$rcx   : 0x0               
$r9    : 0x0               
$rsp   : 0x00007fffffffe590  →  0x0000000000000001
$rip   : 0x00007ffff7fe32b0  →  <_start+0> mov rdi, rsp
$r11   : 0x0               
$r13   : 0x0               
$r15   : 0x0               
$r10   : 0x0               
$r14   : 0x0               
$rax   : 0x0               
$es: 0x00 $ds: 0x00 $fs: 0x00 $cs: 0x33 $ss: 0x2b $gs: 0x00 
```

It's possible that the new code will always return the registers in the same order, but I don't believe that there's any guarantee of this due to `set` being specifically unordered and the order they are returned in does not feel logical to me whilst looking for the contents of a specific register (i.e. I expect $rax to be the first result, I don't expect to have to hunt for it). As such, I believe that this should be reverted so that the behaviour returns results in an expected and consistent order.

### Do you read the docs and look at previously closed issues/PRs for similar cases?

Yes

### Architecture impacted

- [ ] X86
- [X] X64
- [ ] ARM
- [ ] ARM64
- [ ] MIPS
- [ ] MIPS64
- [ ] PPC
- [ ] PPC64
- [ ] RISCV

### Describe your issue. Without a proper reproduction step-by-step, your issue will be ignored.

Run GDB, hit a breakpoint or use `starti`.

Older version show registers in expected order, new version doesn't.

### Minimalist test case

N/A - Any binary should do it

### Additional context?

N/A

### PR DESCRIPTION
**Pull request by therealdreg**
## Description

revert old thread order logic

## Checklist

-  [x] My code follows the code style of this project.
-  [x] My change includes a change to the documentation, if required.
-  [x] If my change adds new code, [adequate tests](docs/testing.md) have been added.
-  [x] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

File: gef.py [hunk 1]
```diff
@@ -6772,11 +6784,11 @@ def do_invoke(self, _: List[str], **kwargs: Any) -> None:
 
         args : argparse.Namespace = kwargs["arguments"]
         if args.registers and args.registers[0]:
-            requested_regs = set(args.registers)
-            valid_regs = set(gef.arch.all_registers) & requested_regs
+            required_regs = set(args.registers)
+            valid_regs = [reg for reg in gef.arch.all_registers if reg in required_regs]
             if valid_regs:
```
Comment (BarryStokes) [inline gef.py:6789]: If the user provided a list of registers, none of which were valid this behaviour will result in it showing _all_ registers as it only updates `regs` (which is set to all above) if there are `valid_regs`. Is this what the user would expect? (Or does the `err` for `invalid_regs` below just stop it showing anything at all?)

File: gef.py [hunk 2]
```diff
@@ -6772,11 +6784,11 @@ def do_invoke(self, _: List[str], **kwargs: Any) -> None:
 
         args : argparse.Namespace = kwargs["arguments"]
         if args.registers and args.registers[0]:
-            requested_regs = set(args.registers)
-            valid_regs = set(gef.arch.all_registers) & requested_regs
+            required_regs = set(args.registers)
```
[the remark this answers is not in the document]
Comment (therealdreg) [reply gef.py:6787]: hey @BarryStokes  I just saw your pull request! I should leave this one just for the threads, and we'll keep yours for registers, agree?

**Comment by Grazfather** [comment]
#989 is for registers, this is for threads.

**Comment by therealdreg** [comment]
> #989 is for registers, this is for threads.

done

File: gef.py [hunk 3]
```diff
@@ -6523,9 +6523,20 @@ def find_tcache() -> int:
...
+            try:
+                tid = int(tid)
+            except ValueError:
+                err(f"Invalid thread id {tid:d}")
+                continue
+            if tid in existing_tids:
+                valid_tids.add(tid)
+            else:
+                err(f"Unknown thread {tid}")
+        return list(valid_tids)
```
Comment (Grazfather) [inline gef.py:6539]: this doesn't even fix the issue, since you are adding to a set.

File: gef.py [hunk 4]
```diff
@@ -6523,9 +6523,20 @@ def find_tcache() -> int:
...
-        existing_tids = set(t.num for t in gdb.selected_inferior().threads())
-        return list(set(tids) & existing_tids)
+        """Check the validity, dedup, and return all valid tids."""
+        existing_tids = [t.num for t in gdb.selected_inferior().threads()]
+        valid_tids = set()
+        for tid in tids:
+            try:
+                tid = int(tid)
+            except ValueError:
+                err(f"Invalid thread id {tid:d}")
```
Comment (Grazfather) [inline gef.py:6533]: `:d` would presumably fail since `int(tid)` failed. Why do we even have to do this?

**Review by Grazfather — CHANGES_REQUESTED** [review]

File: gef.py (hunk 4 above)
Comment (therealdreg) [reply gef.py:6533]: I dont know, I am just reverting to old logic xD

Comment (Grazfather) [reply gef.py:6533]: Please don't just blindly revert.

**Pull request closed without merging**